portal-impl/src/com/liferay/portlet/documentlibrary/service/impl/DLFileEntryLocalServiceImpl.java

```diff
@@ -2771,6 +2790,89 @@ private long _getValidFileEntryTypeId(
 		}
 	}
 
+	private void _inheritRolesPermissions(
+			long parentFolderId, DLFileEntry dlFileEntry)
+		throws PortalException {
+
+		Set<String> commonPermissions = new HashSet<>(
+			Arrays.asList(
+				ActionKeys.DELETE, ActionKeys.PERMISSIONS, ActionKeys.UPDATE,
+				ActionKeys.VIEW));
+
+		long companyId = dlFileEntry.getCompanyId();
+		String parentClassName = DLFolderConstants.getClassName();
+
+		Set<String> dlFolderAllActionIds = SetUtil.fromCollection(
+			ResourceActionsUtil.getModelResourceActions(
+				DLFolderConstants.getClassName()));
+
+		Set<String> dlFileEntryAllActionIds = SetUtil.fromCollection(
+			ResourceActionsUtil.getModelResourceActions(
+				DLFileEntryConstants.getClassName()));
+
+		if (parentFolderId == DLFolderConstants.DEFAULT_PARENT_FOLDER_ID) {
+			int count =
+				_resourcePermissionLocalService.getResourcePermissionsCount(
+					companyId, DLConstants.SERVICE_NAME,
+					ResourceConstants.SCOPE_INDIVIDUAL,
+					String.valueOf(dlFileEntry.getGroupId()));
+
+			if (count == 0) {
+				_resourceLocalService.addResources(
+					companyId, dlFileEntry.getGroupId(), 0,
+					DLConstants.RESOURCE_NAME,
+					String.valueOf(dlFileEntry.getGroupId()), false, true,
+					true);
+			}
+
+			parentClassName = DLConstants.RESOURCE_NAME;
+			parentFolderId = dlFileEntry.getGroupId();
+		}
+
+		Map<Long, Set<String>> dlFolderRoleIdsToActionIds =
+			_resourcePermissionLocalService.
+				getAvailableResourcePermissionActionIds(
+					companyId, parentClassName,
+					ResourceConstants.SCOPE_INDIVIDUAL,
+					String.valueOf(parentFolderId), dlFolderAllActionIds);
+
+		Map<Long, Set<String>> dlFileEntryRoleIdsToActionIds =
+			_resourcePermissionLocalService.
+				getAvailableResourcePermissionActionIds(
+					companyId, DLFileEntryConstants.getClassName(),
+					ResourceConstants.SCOPE_INDIVIDUAL,
+					String.valueOf(dlFileEntry.getFileEntryId()),
+					dlFileEntryAllActionIds);
+
+		Set<Long> dlFolderRoleIds = dlFolderRoleIdsToActionIds.keySet();
+
+		for (Long dlFolderRoleId : dlFolderRoleIds) {
+			Set<String> dlFolderActionIds = dlFolderRoleIdsToActionIds.get(
+				dlFolderRoleId);
+
+			dlFolderActionIds.retainAll(commonPermissions);
+
+			Set<String> dlFileEntryActionIds = dlFolderActionIds;
+
+			if (dlFileEntryRoleIdsToActionIds.containsKey(dlFolderRoleId)) {
+				Set<String> roleIdsToActionIds =
```

Comment:
I see the name which is not correspond with the function. It is maybe change to the different name